### DISCUSSION TITLE
How can we pass dynamic values to a custom rule? can we create a rule with parameters like we create a function with parameters?

### DISCUSSION THREAD

**Discussion by ankur90Git** [microsoft/PSRule#1115] [Q&A] [answered]
How can we pass dynamic values to a custom rule? can we create a rule with parameters like we create a function with parameters?

I have created below rule where I am passing expected value of "$workerRuntime" as hardcoded in the rule. Is there any way I can pass this value dynamically using parameters or any external file?

```powershell
# Synopsis: key FUNCTIONS_WORKER_RUNTIME must have value 'dotnet'
Rule 'FUNCTIONS_WORKER_RUNTIME' {

    $workerRuntime = "dotnet"

    AllOf {
        Exists 'ApplicationSettings.FUNCTIONS_WORKER_RUNTIME' -CaseSensitive
        Within 'ApplicationSettings.FUNCTIONS_WORKER_RUNTIME' $workerRuntime -CaseSensitive
    }
}
```

**Comment by BernieWhite** [comment, marked as answer]
@ankur90Git Thanks for the question.

Currently there isn't a global variable or parameters concept in PSRule, however there is an option to use configuration. Configuration works for PowerShell-based rules but we are looking at YAML and JSON rule support. This should achieve what you are looking for.

You can set configuration key/values in code, file (via `ps-rule.yaml`), or environment variables depending on how you are running PSRule one of these options should work.

See [Options](https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Options/#configuration) for the full details on how to do this.

You could update your rule to be something like:

```powershell
# Synopsis: key FUNCTIONS_WORKER_RUNTIME must have value 'dotnet'
Rule 'FUNCTIONS_WORKER_RUNTIME' {
    AllOf {
        Exists 'ApplicationSettings.FUNCTIONS_WORKER_RUNTIME' -CaseSensitive
        Within 'ApplicationSettings.FUNCTIONS_WORKER_RUNTIME' $configuration.WORKERRUNTIME -CaseSensitive
    }
}
```

Example adding the configuration value to `ps-rule.yaml`:

```yaml
configuration:
  WORKERRUNTIME: 'dotnet'
```

**Reply by ankur90Git** [reply]
Thanks a lot for answering my query.
I tried the solution given by you, just wanted to confirm where we need to provide the path for the yaml configuration file.

I tried using below syntaxes but didn't worked:

1. $option = New-PSRuleOption -Configuration 'C:\Ankur\Powershell_Scripts\Azure\ps-rule.yaml';
    Invoke-PSRule -Path $rulePath -InputPath $inputPath -Option $option;

2. Invoke-PSRule -Path $rulePath -InputPath $inputPath -Option 'C:\Ankur\Powershell_Scripts\Azure\ps-rule.yaml';

**Reply by BernieWhite** [reply]
Your `2.` should work.

1. Would be like this:

```powershell
$option = New-PSRuleOption -Configuration @{ WORKERRUNTIME = 'dotnet' };
Invoke-PSRule -Path $rulePath -InputPath $inputPath -Option $option;
```

or 

```powershell
$option = New-PSRuleOption -Path 'C:\Ankur\Powershell_Scripts\Azure\ps-rule.yaml'
Invoke-PSRule -Path $rulePath -InputPath $inputPath -Option $option;
```

What seems to be the issue with the syntax?
